pkg/radix/tree.go

```diff
@@ -145,29 +130,39 @@ func (t *Tree) Len() int {
 
 // longestPrefix finds the length of the shared prefix
 // of two strings
-func longestPrefix(k1, k2 string) int {
-	max := len(k1)
-	if l := len(k2); l < max {
-		max = l
+func longestPrefix(k1, k2 []byte) int {
+	// for loops can't be inlined, but goto's can. we also use uint to help
+	// out the compiler to prove bounds checks aren't necessary on the index
+	// operations.
+
+	lk1, lk2 := uint(len(k1)), uint(len(k2))
+	i := uint(0)
+
+loop:
+	if lk1 <= i || lk2 <= i {
+		return int(i)
 	}
-	var i int
-	for i = 0; i < max; i++ {
-		if k1[i] != k2[i] {
-			break
-		}
+	if k1[i] != k2[i] {
+		return int(i)
 	}
-	return i
+	i++
+	goto loop
+}
+
+// byteCopy makes a copy of the byte slice.
+func byteCopy(x []byte) []byte {
+	out := make([]byte, len(x))
```

Comment:
I wonder if it would help to allocate a slightly larger slice and hand out sub-slices while there is enough space for it.  When it's too small, create a new slice to slice up further.

Reply:
I see, so automatically do the optimization above. I'll give that a shot.